Evp documentation pass

evqna opened the pull request:
<!--
Thank you for your pull request. Please review these requirements:

Contributors guide: https://github.com/openssl/openssl/blob/master/CONTRIBUTING

Other than that, provide a description above this comment if there isn't one already

If this fixes a github issue, make sure to have a line saying 'Fixes #XXXX' (without quotes) in the commit message.
-->

Contributes documentation to the EVP module, mainly in the form of man page updates to the private key params control operations.

The actual function description is a summary of my understanding of the _pkey__&lt;alg&gt;__ctrl_ implementations and specific algorithm parameter documentation in **genpkey(1)** and **pkeyutl(1)**.

This brings the missing EVP functions docs down to 82 compared to #5193.

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->
- [x] documentation is added or updated

mattcaswell commented on doc/man3/EVP_PKEY_CTX_ctrl.pod:
```diff
@@ -152,20 +254,58 @@ B<nid>. The B<nid> parameter must be B<NID_ffdhe2048>, B<NID_ffdhe3072>,
+
+The EVP_PKEY_CTX_set_ecdh_kdf_type() macro sets the key derivation function type
+to B<kdf> for ECDH key derivation. Possible values are B<EVP_PKEY_ECDH_KDF_NONE>
+and B<EVP_PKEY_ECDH_KDF_X9_62>.
```
This KDF seems particularly badly named! AFAICT this is actually the KDF from X9.63 not X9.62!!! X9.62 is ECDSA but X9.63 is ECDH (the same KDF is also specified in SEC1). If this option is used then the kdf_outlen and md must also be set.

mattcaswell replied:
Perhaps we should create a synonym for this option called `EVP_PKEY_ECDH_KDF_X9_63` and use that in preference in the docs and code.

evqna replied:
Again this appears to be more than just a typo, the implementation in ecdh_kdf is also named ECDH_KDF_X9_62. Are we sure this is an error?

mattcaswell replied:
Yes, I can see it is consistently referred to as the X9.62 KDF throughout the implementation, the API and the original commit message (25af7a5db). However I've looked through X9.62 and I can't see a KDF in it. Comparing the implementation against the X9.63 KDF spec (section 5.6.3) it does seem to match. To be on the safe side I've reached out to the original author (no longer working on the project) to see what he says. I'll let you know.

mattcaswell replied:
The answer I got back on this was:

>The derivation is part of CMS support for ECDH which is defined in RFC5753 et
al. There should be some OIDs in there too which hopefully have X9.63 in them.
It was part of an effort to get DH support for CMS included: I think the only
thing that supported it at the time was Bouncy Castle.
>
>So yes it looks like there is an off by .01 error there.
>
>I can't double check at the moment: I'm away from my main PC and will be for
some days.

mattcaswell commented on doc/man3/EVP_PKEY_CTX_ctrl.pod:
```diff
@@ -152,20 +254,58 @@ B<nid>. The B<nid> parameter must be B<NID_ffdhe2048>, B<NID_ffdhe3072>,
+output length to B<len> for ECDH key derivation.
+
+The EVP_PKEY_CTX_set0_ecdh_kdf_ukm() macro sets the user key material to
+B<ukm> for ECDH key derivation.
```
In X9.63 terms this is the "SharedInfo" parameter

evqna replied:
I could not find any reference to shared info in the X9.63 draft, so I put Shared data instead which seems to serve the same purpose

mattcaswell replied:
It is called SharedInfo in section 5.6.3 (Key Derivation Functions). Make sure you are looking at as a late a draft as possible. Mine is dated January 8th 1999.

mattcaswell commented on doc/man3/EVP_PKEY_CTX_ctrl.pod:
```diff
@@ -152,20 +254,58 @@ B<nid>. The B<nid> parameter must be B<NID_ffdhe2048>, B<NID_ffdhe3072>,
+and B<EVP_PKEY_ECDH_KDF_X9_62>.
+
+The EVP_PKEY_CTX_set_ecdh_kdf_md() macro sets the key derivation function
+message digest to B<md> for ECDH key derivation.
```
X9.63 says this should be SHA1, but SEC1 doesn't seem to prescribe that. We allow any.

evqna commented:
I intentionally did not document the get functions since it seems like all the relevant information is already contained in the set function description and it would only generate noise to repeat that information a second time. What do you think?

Regarding the doc-nits errors, what do I need to do exactly? Should all these macros just be added to private.num or is there something else to do?

mattcaswell commented:
> I intentionally did not document the get functions since it seems like all the relevant information is already contained in the set function description and it would only generate noise to repeat that information a second time. What do you think?

I think we generally do document both the getter and the setter. I'd rather be complete.

> Regarding the doc-nits errors, what do I need to do exactly? Should all these macros just be added to private.num or is there something else to do?

Yes, for each macro that its complaining about you need to add a line in private.num

evqna commented on doc/man3/EVP_PKEY_CTX_ctrl.pod:
```diff
@@ -152,20 +254,58 @@ B<nid>. The B<nid> parameter must be B<NID_ffdhe2048>, B<NID_ffdhe3072>,
 
+The EVP_PKEY_CTX_set_ecdh_cofactor_mode() macro sets the cofactor mode to
+B<cofactor_mode> for ECDH key derivation. Possible values are 1 to enable
+cofactor generation, 0 to disable it and -1 to clear the stored cofactor key.
```
Looking at the code in ec_pmeth it does look like doing a set(-1) does more than just clear the flag, it also frees memory (dctx->co_key) and defaults to the private ec key flags on subsequent get_cofactor_mode() calls. I am having trouble understanding what purpose this -1 parameter actually serves.

mattcaswell replied:
The co_key is just a straight copy of the ec_key with the `EC_FLAG_COFACTOR_ECDH` flag set on it. Presumably so we don't want to make any modifications to the "real" key. We don't want to do anything in this EVP_PKEY_CTX that might affect future operations with this key (just because we want cofactor ECDH now, doesn't mean we will next time). AFAICT there is no real difference between passing 0 and -1 in terms of external behaviour. The former clears the flag on the co_key, and the latter frees the co_key. When it comes to the derivation step we use the co_key if it is present and otherwise use the ec_key, Using the co_key without the `EC_FLAG_COFACTOR_ECDH` flag set will get you exactly the same result as using the ec_key directly.

mattcaswell commented on doc/man3/EVP_PKEY_CTX_ctrl.pod:
```diff
@@ -116,56 +197,182 @@ padding for RSA the algorithm identifier byte is added or checked and removed
+synonymous. When set, the RFC5114 parameters are used instead of new DH
+parameters being generated so all other DH parameters are overriden. Possible
+values are 1, 2 and 3 corresponding to RFC5114 sections 2.1, 2.2 and 2.3.
+0 can be used to clear the stored value. The nid param must have been set.
```
The nid param must *not* have been set. Please note in this para that this is used for parameter generation and that you should create the EVP_PKEY_CTX by using `EVP_PKEY_CTX_new_id(EVP_PKEY_DHX, eng)`. Alternatively I'm wondering whether we should fix `EVP_PKEY_CTX_set_dh_rfc5114` to just use `EVP_PKEY_DH` in the public header since this is obviously a bug...but maybe people are already using it ;-(

evqna replied:
Since the next version is planned to be a major release, wouldn't it make sense to take the opportunity to do this kind of breaking changes?

mattcaswell replied:
> Since the next version is planned to be a major release, wouldn't it make sense to take the opportunity to do this kind of breaking changes?

It will also be a goal of the next release to keep breaking changes to an absolute minimum.

> Also the EVP_PKEY_DHX flag is present both in the header and in the implementation of dh_pkey_paramgen. Maybe it not a bug but intentional behavior?

I would expect to see it in `dh_pkey_paramgen`. The generated parameters are X9.42 style parameters (which correspond to EVP_PKEY_DHX). There is a difference between the type used to create the EVP_PKEY_CTX, and the type of the parameters that are eventually generated using that EVP_PKEY_CTX.

Elsewhere when we generate X9.42 parameters (not using RFC5114), we allow you to create an EVP_PKEY_CTX using `EVP_PKEY_CTX_new_id(EVP_PKEY_DH, eng)`, even though the parameters you are eventually going to generate are `EVP_PKEY_DHX` parameters. But here we don't seem to allow that. It's very inconsistent.

mattcaswell commented on doc/man3/EVP_PKEY_CTX_ctrl.pod:
```diff
@@ -152,20 +254,58 @@ B<nid>. The B<nid> parameter must be B<NID_ffdhe2048>, B<NID_ffdhe3072>,
+
+The EVP_PKEY_CTX_set_dh_kdf_type() macro sets the key derivation function type
+to B<kdf> for DH key derivation. Possible values are B<EVP_PKEY_DH_KDF_NONE>
+and B<EVP_PKEY_DH_KDF_X9_42>.
```
Please can you include some of the above information in the docs here, i.e. that this is the RFC2631 KDF and that if set you must also set the KDF OID, output length and md.

mattcaswell commented on doc/man3/EVP_PKEY_CTX_ctrl.pod:
```diff
@@ -116,56 +197,182 @@ padding for RSA the algorithm identifier byte is added or checked and removed
+The EVP_PKEY_CTX_set0_dh_kdf_oid() macro sets the key derivation function
+object identifier to B<oid> for DH key derivation. The library takes ownership
+of the object identifier so the caller should not free the original memory
+pointed to by B<oid>.
```
Please describe what this OID is for as per my earlier comment

mattcaswell commented on doc/man3/EVP_PKEY_CTX_ctrl.pod:
```diff
@@ -152,20 +254,58 @@ B<nid>. The B<nid> parameter must be B<NID_ffdhe2048>, B<NID_ffdhe3072>,
+object identifier to B<oid> for DH key derivation.
+
+The EVP_PKEY_CTX_set_dh_kdf_md() macro sets the key derivation function
+message digest to B<md> for DH key derivation.
```
Please mention that RFC2631 specifies SHA1 but OpenSSL tolerates other digests.

mattcaswell commented on doc/man3/EVP_PKEY_CTX_ctrl.pod:
```diff
@@ -116,56 +197,182 @@ padding for RSA the algorithm identifier byte is added or checked and removed
+The EVP_PKEY_CTX_set0_dh_kdf_ukm() macro sets the user key material to
+B<ukm> and its length to B<len> for DH key derivation. If B<ukm> is NULL the
+length is set to 0. The library takes ownership of the user key material so
+the caller should not free the original memory pointed to by B<ukm>.
```
Please mention that this is the `partyAInfo` field from RFC2631. I'm not sure what to do about the requirement in that spec that this should be 512 bits.

evqna commented on doc/man3/EVP_PKEY_CTX_ctrl.pod:
```diff
@@ -152,20 +254,58 @@ B<nid>. The B<nid> parameter must be B<NID_ffdhe2048>, B<NID_ffdhe3072>,
+DH parameter options are overriden. 0 can be used to clear the stored value.
+
+The EVP_PKEY_CTX_set_dh_kdf_type() macro sets the key derivation function type
+to B<kdf> for DH key derivation. Possible values are B<EVP_PKEY_DH_KDF_NONE>
```
Is this true only for this specific macro or for all KDF parameters in DH ?

mattcaswell replied:
This is true for all the current KDF macros for DH.

evqna commented on doc/man3/EVP_PKEY_CTX_ctrl.pod:
```diff
@@ -116,56 +197,182 @@ padding for RSA the algorithm identifier byte is added or checked and removed
 B<NID_ffdhe4096>, B<NID_ffdhe6144> or B<NID_ffdhe8192>.  This macro can be
-called during parameter or key generation.
+called during parameter or key generation. The nid must be set before
+the rfc5114 parameter.
```
Indeed I misunderstood this section of the code, thanks for the clarification.

evqna commented on doc/man3/EVP_PKEY_CTX_ctrl.pod:
```diff
@@ -152,20 +254,58 @@ B<nid>. The B<nid> parameter must be B<NID_ffdhe2048>, B<NID_ffdhe3072>,
+and B<EVP_PKEY_DH_KDF_X9_42>.
+
+The EVP_PKEY_CTX_set0_dh_kdf_oid() macro sets the key derivation function
+object identifier to B<oid> for DH key derivation.
```
Does this actually correspond to how OpenSSL uses the OID or is it something the user is supposed to implement himself?
Looking at the code, the usage of this parameter seems to be buried deep into the ASN1 encoding so I couldn't quite figure it out.

mattcaswell replied:
This is used to construct a block of DER which is then hashed to create the shared secret. It has to be the same OID used on both sides otherwise the shared secret that is derived will be different. AFAIK there is no checking to confirm that the eventual CEK is actually for the algorithm defined by the OID. This is something the user is expected to get right, i.e. don't say you are using one algorithm in the OID, but then end up using a different one.

mattcaswell replied:
How about just adding: "This OID should identify the algorithm to be used with the Content Encryption Key."

evqna commented:
Pushed updates addressing most of Matt's comments, except the OID parameter usage (needs more clarification).

mattcaswell commented on doc/man3/EVP_PKEY_CTX_ctrl.pod:
```diff
@@ -116,56 +199,215 @@ padding for RSA the algorithm identifier byte is added or checked and removed
+The library takes ownership of the object identifier so the caller should not
+free the original memory pointed to by B<oid>.
+
+The EVP_PKEY_CTX_get0_dh_kdf_oid() macro gets the key derivation function
```
"gets the key derivation function oid"

evqna commented:
Do you have a link to the latest ANSI X9.63 working draft? I haven't been able to find anything other than the 1998 version.

evqna commented:
Added the suggested OID description and rebased the branch on current master.

As Matt suggested, I added a EVP_PKEY_ECDH_KDF_X9_63 alias in the _ec.h_ header. Is this considered an API change? I guess it is not too important since the next version is planned to be a major release.

mattcaswell commented on include/openssl/ec.h:
```diff
@@ -1457,7 +1457,8 @@ void EC_KEY_METHOD_get_verify(const EC_KEY_METHOD *meth,
 # define EVP_PKEY_ECDH_KDF_NONE                          1
-# define EVP_PKEY_ECDH_KDF_X9_62                         2
+# define EVP_PKEY_ECDH_KDF_X9_63                         2
+# define EVP_PKEY_ECDH_KDF_X9_62   EVP_PKEY_ECDH_KDF_X9_63
```
I'd put a comment here explaining that the original name was an error and is kept here for backwards compatibility reasons. I'd also suggest that if we're going to make this change then we should also update all internal uses of `EVP_PKEY_ECDH_KDF_X9_62` to use `EVP_PKEY_ECDH_KDF_X9_63` instead. For bonus points we should create a new `ECDH_KDF_X9_63()` function that replaces the `ECDH_KDF_X9_62()` function. We'll still need `ECDH_KDF_X9_62()` but it should just call `ECDH_KDF_X9_63()` (again with a comment explaining that the original name was an error). `ECDH_KDF_X9_62()` is an exported public function. I would make `ECDH_KDF_X9_63()` an internal function only.

evqna replied:
What is the reasoning for making the macro identifier public but keeping the function internal? Since everything declared in _ec.h_ is considered public, which header should this new function be declared in?
The original one is used in ec_pmeth and sm2_crypt, both do not seem to share any internal header files.

Also, should internal functions appear in the _libcrypto.num_ file or is it only for exported symbols?

mattcaswell replied:
> What is the reasoning for making the macro identifier public but keeping the function internal?

We don't want to break the API. So anything that was public before must continue to remain public. The use of `EVP_PKEY_ECDH_KDF_X9_63` still seems like a reasonable part of the EVP API. On the other hand the function `ECDH_KDF_X9_62()` completely sidesteps EVP and is part of the low-level ECDH API. The strategic direction for the low-level APIs is to not export them. All new algorithms only get an EVP interface (e.g. the only way to invoke X448 is via the EVP API - you can't use the low-level primitives directly). Therefore my thinking is that `ECDH_KDF_X9_62()` will probably get deprecated at some point - so why create yet another low level function that we don't really want people to use as well? I've not got a strong view on this - so if you feel we should export it then I'm not going to complain too loudly.

> Since everything declared in ec.h is considered public, which header should this new function be declared in?

Probably crypto/include/internal/ec_int.h

> Also, should internal functions appear in the libcrypto.num file or is it only for exported symbols?

It's only for exported symbols. But you shouldn't need to edit it directly. Just invoke "make update" and it should get updated automatically.

mattcaswell replied:
Yes we should start the name with lowercase.

If we add deprecation macros then I don't think we should backport that bit to 1.1.1. Otherwise I think everything else can be backported.

mattcaswell commented:
> As Matt suggested, I added a EVP_PKEY_ECDH_KDF_X9_63 alias in the ec.h header. Is this considered an API change?

We're always allowed to *add* symbols from one release to the next, so this is fine.

evqna commented:
I can squash all commits once the initial round of feedback has been addressed.

Also is worth adding an entry in _CHANGES_ about the doc cleanup and the new KDF define?

romen commented:
> Also is worth adding an entry in CHANGES about the doc cleanup and the new KDF define?

I would suggest one about depreciating *62 and renaming to private *63 methods. 

Not sure about writing a CHANGES entry regarding updated documentation

evqna commented:
What is the procedure to deprecate functionality?

In general for internal macros and functions is the guideline to write documentation at the declaration or the implementation? Looking at some examples in the code, it seems like the standard procedure is to keep it in the headers.

mattcaswell commented:
I think we should squash all the commits, and add the deprecation macros in a separate commit. Then we backport the documentation fixes to 1.1.1, and keep the deprecation macro commit in master only.

romen commented:
> I think we should squash all the commits, and add the deprecation macros in a separate commit. Then we backport the documentation fixes to 1.1.1, and keep the deprecation macro commit in master only.


In light of this I would say that we need 2 CHANGES entries: one about renaming *62 to *63 that is going to be back ported to 1.1.1, and a second one in the "deprecation commit" saying that the old 62 was deprecated. 

I would still like to have the comment about the rationale for *62 being a abi compatibility wrapper around *63 added also where the implementation is: it will make things easier when reviewing that code in the future until we can actually remove the function.

evqna commented:
I forgot to rename the function to the lowercase variant, let me just do that.

evqna commented:
Everything should be good now, thanks for your help.

mattcaswell commented:
> I can take care of merging it if you want.

Great - thanks.

> Regarding merge to 1.1.1 for the first commit: it will definitely require massaging for the CHANGES entry: I could easily do it manually during merge, but I just want to make sure that would be ok, or if that change would warrant a separate PR with separate approval.

Assuming the changes are trivial, then no need for a separate PR.

BTW, kudos to @evqna for this PR. This is a significant contribution to the documentation!!

romen commented:
Merged:
- `master`:
  * ffd8912
  * 9453b19
- `1.1.1-stable`
  * 8710396

Thanks again @evqna for your contribution, I cannot say it with better words than @mattcaswell's :
> kudos to @evqna for this PR. This is a significant contribution to the documentation!!

mspncp commented:
As the creator of issue #5193, I'd like to join in to the laudation of Matt and Nicola :-)

evqna commented:
Thanks everyone!
